### PR TITLE
Implement SPIFFE bundle client and enable istiod xDS authn based on SPIFFE bundles

### PR DESCRIPTION
**Pull request by myidpt**
Enable Istiod to retrieve extra root cert from SPIFFE bundle endpoint for XDS client authn. This can make sure istiod can enforce strict client authn when the client is using a cert issued by a custom CA, as long as a SPIFFE bundle endpoint is offered. In this case, Istiod is still able to verify the cert using the root cert retrieved via the SPIFFE bundle endpoint.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by myidpt** [comment]
@howardjohn Retrieving the root at installation time sounds like a good idea. It does tolerate potential issues due to connection to SPIFFE bundle endpoint.

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -915,6 +917,15 @@ func (s *Server) getRootCertificate(tlsOptions TLSOptions) (*x509.CertPool, erro
 	}
 	cp := x509.NewCertPool()
 	cp.AppendCertsFromPEM(rootCertBytes)
+
+	if features.AuthnSpiffeBundlePath != "" {
```
Comment (costinm) [inline pilot/pkg/bootstrap/server.go:921]: Not sure what is the use case - but let's avoid repeating past mistakes. 

In most cases we need more than one certificate - either make the 'spiffee bundle path' a list, so we can fetch multiple docs, or parse multiple certs from the bundle ( it seems to be only one ). 

Also not entirely sure why this is in pilot - and not pilot-agent ( or at least usable by both ). Is it just so Istiod combines all the roots ( from Secret, config - and Spiffee URLs ) - and distributes them as the config map ?

**Comment by myidpt** [comment]
@costinm I admit we need to support loading multiple root certs *for the workloads*, but this PR is to solve the istiod authenticating xDS client issue.
I'm glad to create follow up PRs to retrieve certs for istio agents. ConfigMap doesn't seem like a good idea. I'd prefer istio-agent directly pulling the bundle from the endpoint (the endpoint can be exposed on Istiod).

File: pilot/pkg/bootstrap/server.go (hunk 1 above)
Comment (myidpt) [reply pilot/pkg/bootstrap/server.go:921]: Sounds reasonable. I'll update the PR tomorrow with ability to load root certs from multiple endpoints.

**Comment by costinm** [comment]
Agreed config map is not ideal - for this kind of use case we need to periodically rotate the root, because it changes. That's also the main reason I'm asking for multiple root support...

Note that config map plus a cron could be used even today - without code changes, and it's very reasonable to expect integration with external CAs to have some CA-specific code running in the cluster - for example as a signer for the K8S API or to distribute the roots.

I'm not asking for agent to pull directly in this PR ( or to pull multiple roots in this PR ), just to make sure we can deal with multiple sources and to have the code in a place where it can be reused by
future PRs that add this to agent. 

Right now Istiod is distributing the roots as a config map - I am not entirely sure if 2 roots are tested.

Keep in mind that some CAs may rotate quite often - in particular given the lack of revocation support. I don't see how we can avoid support for multiple roots.

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -902,7 +903,8 @@ func (s *Server) getCertKeyPaths(tlsOptions TLSOptions) (string, string) {
 	return key, cert
 }
 
-// getRootCertificate returns the root certificate from TLSOptions if available or from ca.
+// getRootCertificate returns the root certificate from TLSOptions if available or from ca, or from SPIFFE bundle
```
Comment (costinm) [inline pilot/pkg/bootstrap/server.go:906]: Plural - or 'the set of root certificates'

We have a pervasive problem of ignoring cert rotation, migration and multiple roots - at least the docs should 
be clear.

Comment (costinm) [reply pilot/pkg/bootstrap/server.go:906]: And the method name could have an 's' at the end - or getRootCertPool - the return value is a set of certificates, indexed by subject and subject key id.

File: pilot/pkg/features/pilot.go [hunk 3]
```diff
@@ -322,4 +322,7 @@ var (
 			"It is safe to disable it if you are quite sure you don't need this feature").Get()
 	InjectionWebhookConfigName = env.RegisterStringVar("INJECTION_WEBHOOK_CONFIG_NAME", "istio-sidecar-injector",
 		"Name of the mutatingwebhookconfiguration to patch, if istioctl is not used.")
+
+	AuthnSpiffeBundlePath = env.RegisterStringVar("AUTHN_SPIFFE_BUNDLE_PATH", "",
```
Comment (costinm) [inline pilot/pkg/features/pilot.go:326]: I would add this to ProxyConfig ( accessible in istiod as MeshConfig.DefaultProxyConfig), as a list of URLs.

File: pkg/spiffe/spiffe.go [hunk 4]
```diff
@@ -90,3 +102,60 @@ func GenCustomSpiffe(identity string) string {
 
 	return URIPrefix + GetTrustDomain() + "/" + identity
 }
+
+// RetrieveSpiffeBundleRootCert retrieves the trusted CA certificate from a SPIFFE bundle endpoint.
+// It can use the system cert pool and the supplied certificates to validate the endpoint.
+func RetrieveSpiffeBundleRootCert(endpoint string, extraTrustedCerts []*x509.Certificate) (*x509.Certificate, error) {
```
Comment (costinm) [inline pkg/spiffe/spiffe.go:108]: 1. Please check the content type of the result - pretty sure the spiffee or JWK define a content-type for the response
2. I'm pretty sure we'll also want to support getting the roots as PEM - for non-spiffee CAs we may support. So 
maybe break this method in 2, one part that handles the response ( after checking content type), and a method that is fetching the response. The second method may move and change to also support retrieving PEM roots, and also could use files in future ( i.e. configmaps).

**Comment by costinm** [comment]
Sorry - just realized my comments contradict each other. I asked to add this to ProxyConfig - and it may be ok as first step, but we already have an official Istio API to configure how to verify client certificates, in Gateway caCertificates.  So the proper API discussion will be if we should on extending Gateway to support a list of URLs where we would fetch caCertificates for MUTUAL client authentication, and possibly DestinationRule with a similar/consistent config (it also has a 'caCertificates' option).

In other words - if we want to support downloading root CAs - either for spifee validation or for non-spiffee (PEM) - we should do it consistently and update the stable APIs, possibly starting with experimental setting in ProxyConfig but with the path to the beta APIs.

**Comment by myidpt** [comment]
@costinm I think I resolved some of your comments. Now the spiffe client can retrieve multiple root certs from different endpoints, and return a map from trustdomain to root certs. With this PR, Istiod isn't able to verify trust domains against the root cert it used for the cert verification. It's not very simple so I will defer it to another PR.

**Comment by myidpt** [comment]
@costinm I'll make a following PR to do the same implementation for istio-agent, so that workloads can load multiple root certs. I'm not sure I fully understand your opinion about the API, but I know the problem. I think that warrants a short design, probably we don't need to resolve in this PR.

**Comment by myidpt** [comment]
@costinm @howardjohn Can you take another look?

File: pilot/pkg/features/pilot.go [hunk 5]
```diff
@@ -322,4 +322,9 @@ var (
 			"It is safe to disable it if you are quite sure you don't need this feature").Get()
 	InjectionWebhookConfigName = env.RegisterStringVar("INJECTION_WEBHOOK_CONFIG_NAME", "istio-sidecar-injector",
 		"Name of the mutatingwebhookconfiguration to patch, if istioctl is not used.")
+
+	SpiffeBundlePaths = env.RegisterStringVar("SPIFFE_BUNDLE_PATHS", "",
+		"The paths to the SPIFFE bundle endpoints, in JSON format. For example: "+
+			"[{\"trustdomain\": \"foo\", \"url\": \"https://exampleurl1.com:123/path\"},"+
```
Comment (costinm) [inline pilot/pkg/features/pilot.go:328]: Ok for the env ( experimental, etc ) - but when we add it to ProxyConfig we should use a simpler format, like:
spiffe_roots:
  trustdomain1: URL1
  trustdomain2: URL2

File: pkg/spiffe/spiffe.go [hunk 6]
```diff
@@ -36,8 +42,21 @@ const (
 var (
 	trustDomain      = defaultTrustDomain
 	trustDomainMutex sync.RWMutex
+
+	spiffeLog = log.RegisterScope("spiffe", "SPIFFE library logging", 0)
 )
 
+type configTuple struct {
```
Comment (costinm) [inline pkg/spiffe/spiffe.go:49]: Map [string]string ?

File: pkg/spiffe/spiffe.go [hunk 7]
```diff
@@ -76,17 +95,104 @@ func GenSpiffeURI(ns, serviceAccount string) (string, error) {
...
 }
+
+// RetrieveSpiffeBundleRootCertsFromStringInput retrieves the trusted CA certificates from a list of SPIFFE bundle endpoints.
+// It can use the system cert pool and the supplied certificates to validate the endpoints.
+// The input endpointTuples should be in the json format of:
+// [
+//		{"trustdomain": "foo", "url": "..."},
+//		{"trustdomain": "bar", "url": "..."}
+// ]
+func RetrieveSpiffeBundleRootCertsFromStringInput(inputString string, extraTrustedCerts []*x509.Certificate) (
```
Comment (costinm) [inline pkg/spiffe/spiffe.go:120]: If you want to do it this way (i.e. bind the cert to a trust domain - which is correct ) - maybe return a map, with key == trustDomain and value == cert ? 

And I assume the trust domain verification will check that the actual cert is right, since TLS verification treats all certs in the pool as equal.

File: pilot/pkg/features/pilot.go (hunk 5 above)
Comment (myidpt) [reply pilot/pkg/features/pilot.go:328]: Yes, I was thinking about this. But it could be confusing if we don't define it well. For example ":" can't be used as the delimiter since it's allowed in URLs.

File: pkg/spiffe/spiffe.go (hunk 7 above)
Comment (myidpt) [reply pkg/spiffe/spiffe.go:120]: Correct. That's exactly what I'm trying to do in followup PRs.

File: pkg/spiffe/spiffe.go (hunk 6 above)
Comment (myidpt) [reply pkg/spiffe/spiffe.go:49]: Yes, that may be simpler.

File: pilot/pkg/bootstrap/server.go [hunk 8]
```diff
@@ -557,7 +558,7 @@ func (s *Server) initDNSTLSListener(dns string, tlsOptions TLSOptions) error {
 	cfg := &tls.Config{
 		GetCertificate: s.getIstiodCertificate,
 		ClientAuth:     tls.NoClientCert,
-		ClientCAs:      root,
+		ClientCAs:      certPool,
```
Comment (JimmyCYJ) [inline pilot/pkg/bootstrap/server.go:561]: If certPool is not nil but ClientAuth==tls.NoClientCert, does Istiod use the cert pool to verify client cert? Could you add a comment about the behavior?

File: pkg/spiffe/spiffe.go [hunk 9]
```diff
@@ -76,17 +94,105 @@ func GenSpiffeURI(ns, serviceAccount string) (string, error) {
...
+		config := &tls.Config{
+			ServerName: u.Hostname(),
+			RootCAs:    caCertPool,
+		}
+
+		httpClient.Transport = &http.Transport{
+			TLSClientConfig: config,
+		}
+
+		resp, err := httpClient.Get(endpoint)
```
Comment (JimmyCYJ) [inline pkg/spiffe/spiffe.go:159]: Do we need a retry here if there is a networking issue that causes the request fail?.

File: pkg/spiffe/spiffe.go [hunk 10]
```diff
@@ -76,17 +94,105 @@ func GenSpiffeURI(ns, serviceAccount string) (string, error) {
...
+	}
+
+	ret := map[string][]*x509.Certificate{}
+	for trustdomain, endpoint := range config.Map {
+		if !strings.HasPrefix(endpoint, "https://") {
+			endpoint = "https://" + endpoint
+		}
+		u, err := url.Parse(endpoint)
+		if err != nil {
+			return nil, fmt.Errorf("failed to split the SPIFFE bundle URL: %v", err)
```
Comment (JimmyCYJ) [inline pkg/spiffe/spiffe.go:147]: Can we add trustdomain and endpoint info into fmt.Errorf()?

File: pkg/spiffe/spiffe.go [hunk 11]
```diff
@@ -76,17 +94,105 @@ func GenSpiffeURI(ns, serviceAccount string) (string, error) {
...
+			RootCAs:    caCertPool,
+		}
+
+		httpClient.Transport = &http.Transport{
+			TLSClientConfig: config,
+		}
+
+		resp, err := httpClient.Get(endpoint)
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch bundle: %v", err)
```
Comment (JimmyCYJ) [inline pkg/spiffe/spiffe.go:161]: Can we add trustdomain and endpoint info into fmt.Errorf()?

File: pkg/spiffe/spiffe.go [hunk 12]
```diff
@@ -76,17 +94,105 @@ func GenSpiffeURI(ns, serviceAccount string) (string, error) {
...
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			b := make([]byte, 1024)
+			n, _ := resp.Body.Read(b)
+			return nil, fmt.Errorf("unexpected status %d fetching bundle: %s", resp.StatusCode, string(b[:n]))
```
Comment (JimmyCYJ) [inline pkg/spiffe/spiffe.go:168]: Can we add trustdomain and endpoint info into fmt.Errorf()?

File: pkg/spiffe/spiffe.go [hunk 13]
```diff
@@ -76,17 +94,105 @@ func GenSpiffeURI(ns, serviceAccount string) (string, error) {
...
+		}
+
+		doc := new(bundleDoc)
+		if err := json.NewDecoder(resp.Body).Decode(doc); err != nil {
+			return nil, fmt.Errorf("failed to decode bundle: %v", err)
```
Comment (JimmyCYJ) [inline pkg/spiffe/spiffe.go:173]: Can we add trustdomain and endpoint info into fmt.Errorf()?

File: pkg/spiffe/spiffe.go [hunk 14]
```diff
@@ -76,17 +94,105 @@ func GenSpiffeURI(ns, serviceAccount string) (string, error) {
...
+		}
+
+		var cert *x509.Certificate
+		for i, key := range doc.Keys {
+			if key.Use == "x509-svid" {
+				if len(key.Certificates) != 1 {
+					return nil, fmt.Errorf("expected a single certificate in x509-svid entry %d; got %d", i, len(key.Certificates))
```
Comment (JimmyCYJ) [inline pkg/spiffe/spiffe.go:180]: Can we add trustdomain and endpoint info into fmt.Errorf()?

File: pkg/spiffe/spiffe_test.go [hunk 15]
```diff
@@ -141,3 +193,98 @@ func TestGenCustomSpiffe(t *testing.T) {
 		}
 	}
 }
+
+func TestRetrieveSpiffeBundleRootCertsFromStringInput(t *testing.T) {
```
Comment (JimmyCYJ) [inline pkg/spiffe/spiffe_test.go:197]: Could you add a brief comment for the test?

Comment (myidpt) [reply pkg/spiffe/spiffe_test.go:221]: Done.
